File: HenzaiFunc.Core/RaytraceGeometry/Hitable.fs
namespace HenzaiFunc.Core.RaytraceGeometry

open HenzaiFunc.Core.Types

type Hitable =

    abstract member HasIntersection: Ray -> bool
    abstract member Intersect: Ray -> bool*LineParameter
    //TODO: Make Point optional
    abstract member IntersectionAcceptable : bool -> LineParameter -> float32 -> Point -> bool
    abstract member NormalForSurfacePoint : Point -> Normal
    abstract member IsObstructedBySelf: Ray -> bool 
    abstract member TMin : float32
    abstract member TMax : float32

File: Henzai.Core/Raytracing/Hitable.cs
using System.Numerics;

namespace Henzai.Core.Raytracing
{
    public interface Hitable
    {
        bool HasIntersection(Ray ray);
        //TODO investigate ValueTuple
        (bool, float) Intersect(Ray ray);
        bool IntersectionAcceptable(bool hasIntersection, float t, float dotView, Vector4 point);
        Vector4 NormalForSurfacePoint(Vector4 point);
        bool IsObstructedBySelf(Ray ray);
        float TMin();
        float TMax();
    }
}